### PR TITLE
bump go-oidc to v3.1.0

### PR DESCRIPTION
**Pull request by aramase**
#### What type of PR is this?
/kind cleanup

#### What this PR does / why we need it:
Bump go-oidc to `v3.1.0`. `v3.1.0` is the minimum required version (contains https://github.com/coreos/go-oidc/pull/315) to support `discoveryURL` in structured authentication configuration.

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

### CODE REVIEW

File: staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/oidc/oidc.go [hunk 1]
```diff
@@ -288,7 +288,7 @@ func New(opts Options) (*Authenticator, error) {
 		// in the authentication configuration.
 		// The go oidc library does not support validating
 		// multiple audiences, so we have to skip the client ID check and do it ourselves.
-		// xref: https://github.com/coreos/go-oidc/issues/397
+		// xref: https://github.com/coreos/go-oidc/v3/oidc/issues/397
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/oidc/oidc.go:291]: heh, revert this bit, the old link was correct

File: staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/oidc/oidc.go [hunk 2]
```diff
@@ -566,7 +566,7 @@ func (v *idTokenVerifier) Verify(ctx context.Context, rawIDToken string) (*oidc.
 }
 
 // verifyAudience verifies the audience field in the ID token matches the expected audience.
-// This is added based on https://github.com/coreos/go-oidc/blob/b203e58c24394ddf5e816706a7645f01280245c7/oidc/verify.go#L275-L281
+// This is added based on https://github.com/coreos/go-oidc/v3/oidc/blob/b203e58c24394ddf5e816706a7645f01280245c7/oidc/verify.go#L275-L281
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/oidc/oidc.go:569]: revert this line, too :)

File: vendor/github.com/coreos/go-oidc/v3/oidc/jwks.go [hunk 3]
```diff
@@ -120,11 +111,10 @@ func (r *remoteKeySet) verify(ctx context.Context, jws *jose.JSONWebSignature) (
...
-	if !r.now().Add(keysExpiryDelta).After(expiry) {
-		// Keys haven't expired, don't refresh.
-		return nil, errors.New("failed to verify id token signature")
-	}
-
+	// If the kid doesn't match, check for new keys from the remote. This is the
+	// strategy recommended by the spec.
+	//
+	// https://openid.net/specs/openid-connect-core-1_0.html#RotateSigKeys
 	keys, err := r.keysFromRemote(ctx)
```
Comment (liggitt) [inline vendor/github.com/coreos/go-oidc/v3/oidc/jwks.go:118]: does this open us up to DOS via random key id in presented tokens?

File: vendor/github.com/coreos/go-oidc/v3/oidc/oidc.go [hunk 4]
```diff
@@ -192,6 +224,16 @@ type UserInfo struct {
...
 }
 
+type userInfoRaw struct {
+	Subject string `json:"sub"`
+	Profile string `json:"profile"`
+	Email   string `json:"email"`
+	// Handle providers that return email_verified as a string
+	// https://forums.aws.amazon.com/thread.jspa?messageID=949441&#949441 and
+	// https://discuss.elastic.co/t/openid-error-after-authenticating-against-aws-cognito/206018/11
+	EmailVerified stringAsBool `json:"email_verified"`
```
Comment (liggitt) [inline vendor/github.com/coreos/go-oidc/v3/oidc/oidc.go:234]: 🤯

File: staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/oidc/oidc.go (hunk 1 above)
Comment (aramase) [reply staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/oidc/oidc.go:291]: find and replace issue :) fixed now!

File: vendor/github.com/coreos/go-oidc/v3/oidc/oidc.go [hunk 5]
```diff
@@ -55,6 +58,36 @@ func ClientContext(ctx context.Context, client *http.Client) context.Context {
...
+//
+//    ctx := oidc.InsecureIssuerURLContext(parentContext, issuerURL)
+//
+//    // Provider will be discovered with the discoveryBaseURL, but use issuerURL
+//    // for future issuer validation.
+//    provider, err := oidc.NewProvider(ctx, discoveryBaseURL)
+//
+// This is insecure because validating the correct issuer is critical for multi-tenant
+// proivders. Any overrides here MUST be carefully reviewed.
+func InsecureIssuerURLContext(ctx context.Context, issuerURL string) context.Context {
```
Comment (liggitt) [inline vendor/github.com/coreos/go-oidc/v3/oidc/oidc.go:87]: this is unused currently, right?

File: vendor/github.com/coreos/go-oidc/v3/oidc/oidc.go [hunk 6]
```diff
@@ -147,13 +179,13 @@ func NewProvider(ctx context.Context, issuer string) (*Provider, error) {
...
 	return &Provider{
-		issuer:       p.Issuer,
+		issuer:       issuerURL,
 		authURL:      p.AuthURL,
 		tokenURL:     p.TokenURL,
 		userInfoURL:  p.UserInfoURL,
 		algorithms:   algs,
 		rawClaims:    body,
-		remoteKeySet: NewRemoteKeySet(ctx, p.JWKSURL),
+		remoteKeySet: NewRemoteKeySet(cloneContext(ctx), p.JWKSURL),
```
Comment (liggitt) [inline vendor/github.com/coreos/go-oidc/v3/oidc/oidc.go:188]: huh, this cloneContext() call means canceling the context passed to NewProvider does nothing to interrupt in-flight requests for background keyset updates ... that's surprising

File: vendor/github.com/coreos/go-oidc/v3/oidc/oidc.go (hunk 4 above)
Comment (liggitt) [reply vendor/github.com/coreos/go-oidc/v3/oidc/oidc.go:234]: I guess we're not making use of the UserInfo fetching at all... still not a huge fan of this relaxing this way

File: vendor/github.com/coreos/go-oidc/v3/oidc/oidc.go (hunk 5 above)
Comment (aramase) [reply vendor/github.com/coreos/go-oidc/v3/oidc/oidc.go:87]: right! i'll wire this up in the next PR with the addition of `discoveryURL` in the config.

Comment (liggitt) [reply vendor/github.com/coreos/go-oidc/v3/oidc/oidc.go:87]: it's a terrifying name to have in our call stack, just trying to understand exactly what it does and how it works

File: vendor/github.com/coreos/go-oidc/v3/oidc/jwks.go (hunk 3 above)
Comment (aramase) [reply vendor/github.com/coreos/go-oidc/v3/oidc/jwks.go:118]: previously there was a 30 second [delta](https://github.com/kubernetes/kubernetes/pull/123402/files#diff-21bdef324de021527b14e3f61ed5085eecd38352385b14a0ab8e8391b1fb1c29L24) but with this change, yes every random kid would result in a call to the jwks endpoint.

File: vendor/github.com/coreos/go-oidc/v3/oidc/oidc.go (hunk 5 above)
Comment (aramase) [reply vendor/github.com/coreos/go-oidc/v3/oidc/oidc.go:87]: It's only used [here](https://github.com/kubernetes/kubernetes/pull/123402/files#diff-06e9f00b62aa58d544d62f2d046009446a72e681c9f26450b15fafcb0ffd8ce6R168-R172)


```go
ctx := oidc.InsecureIssuerURLContext(parentContext, opts.JWTAuthenticator.Issuer.URL)
provider, err := oidc.NewProvider(ctx, opts.JWTAuthenticator.Issuer.DiscoveryURL) // if discovery url is not empty
...
```

File: vendor/github.com/coreos/go-oidc/v3/oidc/jwks.go (hunk 3 above)
Comment (aramase) [reply vendor/github.com/coreos/go-oidc/v3/oidc/jwks.go:118]: Chatted with @enj offline and he suggested avoiding the dep bump and instead setting `SkipIssuerCheck=true`, performing the issuer validation in authenticator (similar to audience check) when discovery URL is set. I'll make that change in a separate PR and close this after.

Comment (aramase) [reply vendor/github.com/coreos/go-oidc/v3/oidc/jwks.go:118]: opened https://github.com/kubernetes/kubernetes/pull/123527

**Comment by aramase** [comment]
Implemented the changes in https://github.com/kubernetes/kubernetes/pull/123527 to support `discoveryURL` and this bump is no longer required.

/close

**Pull request closed without merging**